commit b2450ba47cccccacc20c8edb887e28790f2549f1
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2019-06-13T21:03:26+00:00 (4 months ago)
    
    MANTA-4307 Make boray, electric-boray, and buckets-postgres proper Manta services

diff --git a/config/services/boray/service.json b/config/services/boray/service.json
new file mode 100644
index 0000000..2160cbc
--- /dev/null
+++ b/config/services/boray/service.json
@@ -0,0 +1,9 @@
+{
+	"params": {
+		"networks": [ "manta", "admin" ],
+		"ram": 256
+	},
+	"metadata": {
+		"MORAY_MAX_PG_CONNS": 16
+	}
+}
diff --git a/config/services/boray/service.json.lab b/config/services/boray/service.json.lab
new file mode 100644
index 0000000..86322f7
--- /dev/null
+++ b/config/services/boray/service.json.lab
@@ -0,0 +1,5 @@
+{
+	"params": {
+		"ram": 1024
+	}
+}
diff --git a/config/services/boray/service.json.production b/config/services/boray/service.json.production
new file mode 100644
index 0000000..4504983
--- /dev/null
+++ b/config/services/boray/service.json.production
@@ -0,0 +1,6 @@
+{
+	"params": {
+		"ram": 32768,
+                "quota": 100
+	}
+}
diff --git a/config/services/buckets-postgres/service.json b/config/services/buckets-postgres/service.json
new file mode 100644
index 0000000..8a347c3
--- /dev/null
+++ b/config/services/buckets-postgres/service.json
@@ -0,0 +1,10 @@
+{
+	"params": {
+		"networks": [ "manta", "admin" ],
+		"ram": 2048,
+		"max_lwps": 256000
+	},
+	"metadata": {
+		"PG_DIR": "/manatee/pg/data"
+	}
+}
diff --git a/config/services/buckets-postgres/service.json.lab b/config/services/buckets-postgres/service.json.lab
new file mode 100644
index 0000000..519cc8a
--- /dev/null
+++ b/config/services/buckets-postgres/service.json.lab
@@ -0,0 +1,6 @@
+{
+	"params": {
+                "quota": 30
+	}
+}
+
diff --git a/config/services/buckets-postgres/service.json.production b/config/services/buckets-postgres/service.json.production
new file mode 100644
index 0000000..d7e80c9
--- /dev/null
+++ b/config/services/buckets-postgres/service.json.production
@@ -0,0 +1,6 @@
+{
+	"params": {
+		"ram": 32768,
+                "quota": 4000
+	}
+}
diff --git a/config/services/electric-boray/service.json b/config/services/electric-boray/service.json
new file mode 100644
index 0000000..ce40041
--- /dev/null
+++ b/config/services/electric-boray/service.json
@@ -0,0 +1,7 @@
+{
+	"params": {
+		"networks": [ "manta", "admin" ],
+		"ram": 2048,
+		"mdata_exec_timeout": 3600
+	}
+}
diff --git a/config/services/electric-boray/service.json.lab b/config/services/electric-boray/service.json.lab
new file mode 100644
index 0000000..49a2fab
--- /dev/null
+++ b/config/services/electric-boray/service.json.lab
@@ -0,0 +1,5 @@
+{
+	"params": {
+		"ram": 3072
+	}
+}
diff --git a/config/services/electric-boray/service.json.production b/config/services/electric-boray/service.json.production
new file mode 100644
index 0000000..4d57911
--- /dev/null
+++ b/config/services/electric-boray/service.json.production
@@ -0,0 +1,6 @@
+{
+	"params": {
+		"ram": 16384,
+                "quota": 100
+	}
+}
diff --git a/docs/man/man1/manta-adm.md b/docs/man/man1/manta-adm.md
index 65bced0..5cf829d 100644
--- a/docs/man/man1/manta-adm.md
+++ b/docs/man/man1/manta-adm.md
@@ -64,6 +64,15 @@ Services that are part of the Manta application include:
 **authcache**
   Stores user identity information
 
+**boray**
+  PostgreSQL interface for manta buckets system
+
+**buckets-postgres**
+  PostgreSQL databases used for storing object metadata for the buckets system
+
+**electric-boray**
+  Boray proxy for buckets that handles sharding using consistent hashing
+
 **electric-moray**
   Moray proxy that handles sharding using consistent hashing
 
@@ -695,7 +704,7 @@ See above for information on the `-l`, `-H`, and `-o` options.
 
 `-j, --json`
   Instead of the default text-based output, emit a JSON representation of the
-  summary information reported with the `-s/--summary` command.  This format is 
+  summary information reported with the `-s/--summary` command.  This format is
   suitable for use with `manta-adm update`.  This option cannot be combined with
   `-c/--bycn`, `-a/--all`, `-H/--omit-header`, or `-o/--columns`, and it _must_
   be combined with `-s/--summary`.  (Future versions of this command may support
diff --git a/lib/deploy.js b/lib/deploy.js
index 59bbc83..d375926 100644
--- a/lib/deploy.js
+++ b/lib/deploy.js
@@ -22,6 +22,7 @@ var util = require('util');
 var vasync = require('vasync');
 var sprintf = util.format;
 var common = require('./common');
+var services = require('./services');
 var mod_ssl = require('./ssl');
 var EventEmitter = require('events').EventEmitter;
 var VError = require('verror').VError;
@@ -973,7 +974,7 @@ function createInstance(self, app, svc, cb) {
 	/*
 	 * Prefix with the shard for things that are shardable...
 	 */
-	if (['postgres', 'moray'].indexOf(self.svcname) !== -1) {
+	if (services.serviceIsSharded(self.svcname)) {
 		params.alias = shard + '.' + params.alias;
 	}
 
@@ -1016,6 +1017,18 @@ function createInstance(self, app, svc, cb) {
 		    shard, service_root);
 	}
 
+	if (self.svcname === 'buckets-postgres') {
+		metadata.SERVICE_NAME = sprintf('%s.boray.%s',
+		    shard, service_root);
+		metadata.MANATEE_SHARD_PATH = sprintf('/manatee/%s',
+		    metadata.SERVICE_NAME);
+	}
+
+	if (self.svcname === 'boray') {
+		metadata.SERVICE_NAME = sprintf('%s.boray.%s',
+		    shard, service_root);
+	}
+
 	if (self.svcname === 'storage') {
 		metadata.SERVICE_NAME = sprintf('stor.%s', service_root);
 	}
diff --git a/lib/layout.js b/lib/layout.js
index 5d62e8e..b6c4484 100644
--- a/lib/layout.js
+++ b/lib/layout.js
@@ -115,7 +115,9 @@ var ML_SERVICES_EXACT = {
 var ML_NPERSHARD_INSTANCES = 3;
 var ML_SERVICES_PER_SHARD = {
     'postgres': ML_NPERSHARD_INSTANCES,
-    'moray': ML_NPERSHARD_INSTANCES
+    'moray': ML_NPERSHARD_INSTANCES,
+    'buckets-postgres': ML_NPERSHARD_INSTANCES,
+    'boray': ML_NPERSHARD_INSTANCES
 };
 
 /*
@@ -130,7 +132,8 @@ var ML_FRONTDOOR_RATIOS = {
     'authcache': 1,
     'electric-moray': ML_FRONTDOOR_NMAXINSTANCES,
     'webapi': ML_FRONTDOOR_NMAXINSTANCES,
-    'loadbalancer': ML_FRONTDOOR_NMAXINSTANCES
+    'loadbalancer': ML_FRONTDOOR_NMAXINSTANCES,
+    'electric-boray': ML_FRONTDOOR_NMAXINSTANCES
 };
 
 /*
diff --git a/lib/services.js b/lib/services.js
index e965e92..36ff6dc 100644
--- a/lib/services.js
+++ b/lib/services.js
@@ -54,7 +54,10 @@ var mSvcNames = [
     'reshard',
     'pgstatsmon',
     'garbage-collector',
-    'prometheus'
+    'prometheus',
+    'buckets-postgres',
+    'boray',
+    'electric-boray'
 ];
 
 /*
@@ -80,7 +83,10 @@ var mSvcConfigs = {
     'pgstatsmon':	{ 'oneach': true,  'probes': true,  'sharded': false },
     'garbage-collector': { 'oneach': true, 'probes': true,  'sharded': false },
     'prometheus':	{ 'oneach': true,  'probes': true,  'sharded': false },
-    'marlin':		{ 'oneach': false, 'probes': false, 'sharded': false }
+    'marlin':		{ 'oneach': false, 'probes': false, 'sharded': false },
+    'buckets-postgres':	{ 'oneach': true,  'probes': true,  'sharded': true  },
+    'boray':		{ 'oneach': true,  'probes': true,  'sharded': true  },
+    'electric-boray':	{ 'oneach': true,  'probes': true,  'sharded': false }
 };
 
 exports.mSvcNames = mSvcNames;
diff --git a/man/man1/manta-adm.1 b/man/man1/manta-adm.1
index 43968fb..732dd6b 100644
--- a/man/man1/manta-adm.1
+++ b/man/man1/manta-adm.1
@@ -60,6 +60,15 @@ Services that are part of the Manta application include:
 \fBauthcache\fP
 Stores user identity information
 .TP
+\fBboray\fP
+PostgreSQL interface for manta buckets system
+.TP
+\fBbuckets\-postgres\fP
+PostgreSQL databases used for storing object metadata for the buckets system
+.TP
+\fBelectric\-boray\fP
+Boray proxy for buckets that handles sharding using consistent hashing
+.TP
 \fBelectric\-moray\fP
 Moray proxy that handles sharding using consistent hashing
 .TP
@@ -103,7 +112,7 @@ PostgreSQL databases used for storing object and job metadata
 Monitoring system for Postgres
 .TP
 \fBprometheus\fP
-Time-series database that aggregates Manta and Triton metrics
+Time\-series database that aggregates Manta and Triton metrics
 .TP
 \fBreshard\fP
 Automated resharding system for Manta
@@ -739,7 +748,7 @@ logically apply).  The count for each group is also reported.  With
 .TP
 \fB\fC\-j, \-\-json\fR
 Instead of the default text\-based output, emit a JSON representation of the
-summary information reported with the \fB\fC\-s/\-\-summary\fR command.  This format is 
+summary information reported with the \fB\fC\-s/\-\-summary\fR command.  This format is
 suitable for use with \fB\fCmanta\-adm update\fR\&.  This option cannot be combined with
 \fB\fC\-c/\-\-bycn\fR, \fB\fC\-a/\-\-all\fR, \fB\fC\-H/\-\-omit\-header\fR, or \fB\fC\-o/\-\-columns\fR, and it \fImust\fP
 be combined with \fB\fC\-s/\-\-summary\fR\&.  (Future versions of this command may support
diff --git a/test/tst.layout.js b/test/tst.layout.js
index 70b0161..39fcbb7 100644
--- a/test/tst.layout.js
+++ b/test/tst.layout.js
@@ -347,7 +347,10 @@ var images = {
     'ops': 'OPS_IMAGE0',
     'madtom': 'MADTOM_IMAGE0',
     'marlin-dashboard': 'DASHBOARD_IMAGE0',
-    'marlin': 'MARLIN_IMAGE0'
+    'marlin': 'MARLIN_IMAGE0',
+    'buckets-postgres': 'BUCKETS_POSTGRES_IMAGE0',
+    'boray': 'BORAY_IMAGE0',
+    'electric-boray': 'ELECTRIC_BORAY_IMAGE0'
 };
 
 var nrun = 0;
diff --git a/test/tst.layout.js.out b/test/tst.layout.js.out
index 5bdf498..4a3c2ae 100644
--- a/test/tst.layout.js.out
+++ b/test/tst.layout.js.out
@@ -57,6 +57,31 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            },
+            "2": {
+                "BORAY_IMAGE0": 1
+            },
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -105,6 +130,31 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            },
+            "2": {
+                "BORAY_IMAGE0": 1
+            },
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -150,6 +200,31 @@ generated config:
         }
     },
     "server_r02_metadata00": {
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            },
+            "2": {
+                "BORAY_IMAGE0": 1
+            },
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -232,12 +307,19 @@ summary:
      pgstatsmon           -                 
      garbage-collector     -                 
      prometheus           -                 
+     electric-boray       -                3
      postgres             1                3
      postgres             2                3
      postgres             3                3
      moray                1                3
      moray                2                3
      moray                3                3
+     buckets-postgres     1                3
+     buckets-postgres     2                3
+     buckets-postgres     3                3
+     boray                1                3
+     boray                2                3
+     boray                3                3
 --------------------------------------------------
 --------------------------------------------------
 test case: invalid config: missing value (nshards)
@@ -544,6 +626,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 2
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 3
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 3
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 2
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 2
         },
@@ -616,8 +711,11 @@ summary:
      pgstatsmon           -                 
      garbage-collector     -                 
      prometheus           -                 
+     electric-boray       -                2
      postgres             1                3
      moray                1                3
+     buckets-postgres     1                3
+     boray                1                3
 --------------------------------------------------
 --------------------------------------------------
 test case: trivial two-system case, 3 shards
@@ -645,6 +743,31 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 2
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 3
+            },
+            "2": {
+                "BORAY_IMAGE0": 3
+            },
+            "3": {
+                "BORAY_IMAGE0": 3
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 3
+            },
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 3
+            },
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 3
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 2
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 2
         },
@@ -729,12 +852,19 @@ summary:
      pgstatsmon           -                 
      garbage-collector     -                 
      prometheus           -                 
+     electric-boray       -                2
      postgres             1                3
      postgres             2                3
      postgres             3                3
      moray                1                3
      moray                2                3
      moray                3                3
+     buckets-postgres     1                3
+     buckets-postgres     2                3
+     buckets-postgres     3                3
+     boray                1                3
+     boray                2                3
+     boray                3                3
 --------------------------------------------------
 --------------------------------------------------
 test case: one rack with fewer servers
@@ -816,6 +946,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -843,6 +986,19 @@ generated config:
         }
     },
     "server_r00_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -870,6 +1026,9 @@ generated config:
         }
     },
     "server_r00_metadata02": {
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -887,6 +1046,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -914,6 +1086,19 @@ generated config:
         }
     },
     "server_r01_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -938,6 +1123,9 @@ generated config:
         }
     },
     "server_r01_metadata02": {
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -952,6 +1140,19 @@ generated config:
         }
     },
     "server_r02_metadata00": {
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -979,6 +1180,19 @@ generated config:
         }
     },
     "server_r02_metadata02": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1047,10 +1261,15 @@ summary:
      pgstatsmon           -                 
      garbage-collector     -                 
      prometheus           -                 
+     electric-boray       -                8
      postgres             1                3
      postgres             2                3
      moray                1                3
      moray                2                3
+     buckets-postgres     1                3
+     buckets-postgres     2                3
+     boray                1                3
+     boray                2                3
 --------------------------------------------------
 --------------------------------------------------
 test case: 3-rack, 4-shard deployment
@@ -1174,6 +1393,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1198,6 +1430,19 @@ generated config:
         }
     },
     "server_r00_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1222,6 +1467,19 @@ generated config:
         }
     },
     "server_r00_metadata02": {
+        "boray": {
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1246,6 +1504,19 @@ generated config:
         }
     },
     "server_r00_metadata03": {
+        "boray": {
+            "4": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "4": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1273,6 +1544,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1297,6 +1581,19 @@ generated config:
         }
     },
     "server_r01_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1321,6 +1618,19 @@ generated config:
         }
     },
     "server_r01_metadata02": {
+        "boray": {
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1345,6 +1655,19 @@ generated config:
         }
     },
     "server_r01_metadata03": {
+        "boray": {
+            "4": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "4": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1369,6 +1692,19 @@ generated config:
         }
     },
     "server_r02_metadata00": {
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1393,6 +1729,19 @@ generated config:
         }
     },
     "server_r02_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1417,6 +1766,19 @@ generated config:
         }
     },
     "server_r02_metadata02": {
+        "boray": {
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1441,6 +1803,19 @@ generated config:
         }
     },
     "server_r02_metadata03": {
+        "boray": {
+            "4": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "4": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1533,6 +1908,7 @@ summary:
      pgstatsmon           -                 
      garbage-collector     -                 
      prometheus           -                 
+     electric-boray       -               12
      postgres             1                3
      postgres             2                3
      postgres             3                3
@@ -1541,6 +1917,14 @@ summary:
      moray                2                3
      moray                3                3
      moray                4                3
+     buckets-postgres     1                3
+     buckets-postgres     2                3
+     buckets-postgres     3                3
+     buckets-postgres     4                3
+     boray                1                3
+     boray                2                3
+     boray                3                3
+     boray                4                3
 --------------------------------------------------
 --------------------------------------------------
 test case: 3-AZ, 1-rack-per-AZ, 1-shard deployment
@@ -1598,6 +1982,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1644,6 +2041,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1687,6 +2097,19 @@ generated config:
 }
 {
     "az2_server_r00_metadata00": {
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1748,8 +2171,11 @@ summary:
      pgstatsmon           -                                                   
      garbage-collector     -                                                   
      prometheus           -                                                   
+     electric-boray       -                1                1                1
      postgres             1                1                1                1
      moray                1                1                1                1
+     buckets-postgres     1                1                1                1
+     boray                1                1                1                1
 --------------------------------------------------
 --------------------------------------------------
 test case: 3-AZ, 1-rack-per-AZ, 2-shard deployment
@@ -1807,6 +2233,25 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            },
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1859,6 +2304,25 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            },
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1908,6 +2372,25 @@ generated config:
 }
 {
     "az2_server_r00_metadata00": {
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            },
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            },
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -1976,10 +2459,15 @@ summary:
      pgstatsmon           -                                                   
      garbage-collector     -                                                   
      prometheus           -                                                   
+     electric-boray       -                1                1                1
      postgres             1                1                1                1
      postgres             2                1                1                1
      moray                1                1                1                1
      moray                2                1                1                1
+     buckets-postgres     1                1                1                1
+     buckets-postgres     2                1                1                1
+     boray                1                1                1                1
+     boray                2                1                1                1
 --------------------------------------------------
 --------------------------------------------------
 test case: 3-AZ, 2-racks-per-AZ, 3-shard deployment
@@ -2121,6 +2609,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2148,6 +2649,19 @@ generated config:
         }
     },
     "az0_server_r00_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2172,6 +2686,19 @@ generated config:
         }
     },
     "az0_server_r00_metadata02": {
+        "boray": {
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2225,6 +2752,19 @@ generated config:
         "authcache": {
             "AUTHCACHE_IMAGE0": 1
         },
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2252,6 +2792,19 @@ generated config:
         }
     },
     "az1_server_r00_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2276,6 +2829,19 @@ generated config:
         }
     },
     "az1_server_r00_metadata02": {
+        "boray": {
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2326,6 +2892,19 @@ generated config:
 }
 {
     "az2_server_r00_metadata00": {
+        "boray": {
+            "1": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "1": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2353,6 +2932,19 @@ generated config:
         }
     },
     "az2_server_r00_metadata01": {
+        "boray": {
+            "2": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "2": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2377,6 +2969,19 @@ generated config:
         }
     },
     "az2_server_r00_metadata02": {
+        "boray": {
+            "3": {
+                "BORAY_IMAGE0": 1
+            }
+        },
+        "buckets-postgres": {
+            "3": {
+                "BUCKETS_POSTGRES_IMAGE0": 1
+            }
+        },
+        "electric-boray": {
+            "ELECTRIC_BORAY_IMAGE0": 1
+        },
         "electric-moray": {
             "ELECTRIC_MORAY_IMAGE0": 1
         },
@@ -2445,10 +3050,17 @@ summary:
      pgstatsmon           -                                                   
      garbage-collector     -                                                   
      prometheus           -                                                   
+     electric-boray       -                3                3                3
      postgres             1                1                1                1
      postgres             2                1                1                1
      postgres             3                1                1                1
      moray                1                1                1                1
      moray                2                1                1                1
      moray                3                1                1                1
+     buckets-postgres     1                1                1                1
+     buckets-postgres     2                1                1                1
+     buckets-postgres     3                1                1                1
+     boray                1                1                1                1
+     boray                2                1                1                1
+     boray                3                1                1                1
 --------------------------------------------------
diff --git a/test/tst.services.js b/test/tst.services.js
index 12da351..ac0bfe2 100644
--- a/test/tst.services.js
+++ b/test/tst.services.js
@@ -41,7 +41,10 @@ var knownServices = [
     'reshard',
     'pgstatsmon',
     'garbage-collector',
-    'prometheus'
+    'prometheus',
+    'buckets-postgres',
+    'boray',
+    'electric-boray'
 ];
 
 function main()
@@ -67,7 +70,8 @@ function main()
 	 * Test serviceIsSharded().
 	 */
 	sharded = knownServices.filter(services.serviceIsSharded).sort();
-	assertplus.deepEqual([ 'moray', 'postgres' ], sharded);
+	assertplus.deepEqual(
+	    [ 'boray', 'buckets-postgres', 'moray', 'postgres' ], sharded);
 
 	/*
 	 * Test serviceSupportsOneach().
